### PR TITLE
LRAC-7624 ACTimeFilter Macro

### PR DESCRIPTION
**Pull request by ctang3**

### CODE REVIEW

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+<body>
+<table cellpadding="1" cellspacing="1" border="1">
+  <thead>
+  <tr><td rowspan="1" colspan="3">ACTimeFilter</td></tr>
+  </thead>
+  <tbody>
+  <!-- TIME FILTER ELEMENTS  -->
+  <tr>
+    <td>TIME_FILTER_DROPDOWN</td>
+    <td>//div[@class='dropdown dropdown-range-key-root']/button</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:14]: Hey @ctang3 , just a suggestion in order to let your selector more robust, I would build it using a `contains` pointing to one of the multiple classes available, so if a class is added/removed, your selector will have less chance of being impacted by these changes.

So the selector would be something like:

`//div[contains(@class,'dropdown-range-key-root')]/button`

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    <td></td>
+  </tr>
+  <tr>
+    <td>CARD_TIME_FILTER_DROPDOWN</td>
+    <td>(//h5[@class='card-title' and text()='${key_cardName}']/following::div[@class='dropdown dropdown-range-key-root']/button)[1]</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:19]: I would refactor the last part of this selector using the contains strategy above.

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    <td></td>
+  </tr>
+  <tr>
+    <td>TIME_FILTER_DROPDOWN_OPTION</td>
+    <td>//div[@class='dropdown-menu dropdown-range-key-menu-root show']//button[text()='${key_dropdownOption}']</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>TIME_FILTER_DROPDOWN_MENU</td>
+    <td>//div[@class='dropdown-menu dropdown-range-key-menu-root show']</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:29]: I would refactor using the _contains strategy_ like:

`//div[contains(@class,'dropdown-menu') and contains (@class,'show')]`
or 
`//div[contains(@class,'dropdown') and contains (@class,'show')]`

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    <td></td>
+  </tr>
+  <tr>
+    <td>DATE_PICKER_DROPDOWN_OPTION</td>
+    <td>//TODO</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>DATE_PICKER_DAY</td>
+    <td>//table[@class='calendar-root']//button[@class='button-root btn btn-unstyled day-root' and text()='${key_day}']"</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:54]: I would refactor using the _contains strategy_ like:

//table[@class='calendar-root']//button[contains(@class,'day') and text()='30']

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    <td>//div[@class='date-picker-root']</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>DATE_PICKER_MONTH_DROPDOWN</td>
+    <td>//select[@class='date-picker-select-root form-control'][1]</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>DATE_PICKER_YEAR_DROPDOWN</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:43]: I would refactor using the _contains strategy_ like:

`//select[contains(@class,'date-picker-select-root')][1]`

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    <td>//select[@class='date-picker-select-root form-control'][2]</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:44]: I would refactor using the _contains strategy_ like:

`//select[contains(@class,'date-picker-select-root')][2]`

File: portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    <td>//div[@class='dropdown dropdown-range-key-root']/button</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>CARD_TIME_FILTER_DROPDOWN</td>
+    <td>(//h5[@class='card-title' and text()='${key_cardName}']/following::div[@class='dropdown dropdown-range-key-root']/button)[1]</td>
+    <td></td>
+  </tr>
+  <tr>
+    <td>TIME_FILTER_DROPDOWN_OPTION</td>
```
Comment (alexcandidos) [inline portal-web/test/functional/com/liferay/portalweb/paths/analyticscloud/ACTimeFilter.path:23]: I would refactor using the _contains strategy_ like:

`//div[contains(@class,'dropdown-menu') and contains (@class,'show')]//button[text()='More Preset Periods']`

**Review by alexcandidos — COMMENTED** [review]
Hey @ctang3 , your code looks good to me!! 💯 

I just added a few comments with some improvement suggestions in your selectors.

**Comment by ctang3** [comment]
Updated the selectors to follow your suggestions ✅ 

This is a great idea to use `contains()` more frequently to reduce how often we need to update XPaths. With this strategy we just need to keep in mind to not overdo it and be too fuzzy with the selectors that are shared and check various places for them across various AC pages depending on what the selector is for. Or if something that is within the contains() gets removed/renamed (although its a little more likely for something to be added than removed).

**Comment by ctang3** [comment]
ci:forward

**Pull request closed without merging**